datashader/transfer_functions/__init__.py

```diff
@@ -357,19 +357,31 @@ def _colorize(agg, color_key, how, span, min_alpha, name):
     # min/max of the data
     if span is None:
         span = [np.nanmin(total).item(), np.nanmax(total).item()]
+        offset = span[0]
         if how != 'eq_hist':
             span = _normalize_interpolate_how(how)([0, span[1] - span[0]], 0)
+        # If there are elements that are zero or lower, they should be fully
+        # transparent and masked out
```

Comment:
I'm having trouble following this logic. Negative values in the original data are valid for displaying, so why would they be masked out here?

Reply:
As we were just discussing, maybe this was written without taking into account the fact that categorical aggregates can now be performed with any reduction, including mean, sum etc and can therefore be negative. Maybe we need some way to indicate counts?

Reply:
@jbednar @philippjfr in my primary use case the issue wasn't the new reductions, but rather using `_colorize` with a fixed span (which is also a new feature).  If I set span to sometime like (10, 100) and I have pixel with a count of 5 the old code would subtract 10 and arrive at -5.  I believe the correct behavior is similar to how fixed spans behave in other parts of the code.  All non-zero numbers are clipped to a range 10-100.

You can see these issues by checking out commit `bea6f40` which has the unittest showing the issue, while the following commit fixes the issue.  In the unittests I tests various scenarios where the fixed spans overlap the actual data in different ways and then verify that the alpha channels are calculated as expected.

Reply:
Sorry, I see what you are asking about now, my bad!

Yes this was protective code to deal with incoming data that had negative values.  I'll write a unittest for this case to demonstrate the issue.  Sorry for both the initial confusion and not including a test the first time.

Reply:
Okay, I have pushed a test up, but while implementing it really made me question what the correct behavior is (as I think are as well)

The [original code](https://github.com/holoviz/datashader/blob/7c34b668508deb04dffd7a3db076faa618138da6/datashader/transfer_functions/__init__.py#L358) prior to categorical aggregations supporting sum etc. had this code:

```
    if offset == 0:
        mask = mask | (total <= 0)
```

Since the second line had `total <= 0` I took that to mean that now that offset could be negative, the intention was to mask out negative values as well.  Anyways, I spent some time trying to think of how to get the desired behavior in all cases but haven't come across anything.  I'm happy to implement whatever you think best...or treat this question as a separate PR/feature.

Reply:
Thanks for your efforts on this. I'll think about it a bit and get back to you. Would be great to get this into the imminent 0.11.0 release.

Reply:
Yes; this section of the codebase is a mix of code that was written assuming colorizing would only be done with counts (and thus always non-negative), and other parts that were updated.  So it's tough to reason about it and detect all the places that need to be updated so that they work properly with the new `by()` support.  I think the changes made in this PR so far definitely need to be made in some form, but they need to go further before they actually match the full range of what colorize can now be asked to render.

Reply:
@jbednar @philippjfr as it relates to the 0.11.0 release.  This PR should still be identical behavior to the introduction of `by()` and `span` support and it fixes a fundamental issue with `span` that was in my PR #875.  In other words, I don't think this PR breaks anything that wasn't already broken and fixes something that should work.

Reply:
Ok, I'll merge this now, and then work on master to try to do something about negative values.